ghcide/src/Text/Fuzzy/Parallel.hs

```diff
@@ -1,96 +1,90 @@
 -- | Parallel versions of 'filter' and 'simpleFilter'
+
 module Text.Fuzzy.Parallel
 (   filter,
     simpleFilter,
-    Scored(..),
-    -- reexports
-    Fuzzy,
+    Scored(..)
 ) where
 
-import           Control.Monad.ST            (runST)
-import           Control.Parallel.Strategies (Eval, Strategy, evalTraversable,
-                                              parTraversable, rseq, using)
-import           Data.Monoid.Textual         (TextualMonoid)
-import           Data.Vector                 (Vector, (!))
-import qualified Data.Vector                 as V
--- need to use a stable sort
-import           Data.Bifunctor              (second)
-import           Data.Char                   (toLower)
-import           Data.Maybe                  (fromMaybe)
-import qualified Data.Monoid.Textual         as T
+import           Control.Parallel.Strategies (rseq, using, parList, evalList)
+import           Data.Bits                   ((.|.))
+import           Data.Maybe                  (fromMaybe, mapMaybe)
+import qualified Data.Text                   as T
+import qualified Data.Text.Internal          as T
+import qualified Data.Text.Array             as TA
 import           Prelude                     hiding (filter)
-import           Text.Fuzzy                  (Fuzzy (..))
 
-data Scored a = Scored {score_ :: !Int, original:: !a}
-  deriving (Functor,Show)
+data Scored a = Scored {score :: !Int, original:: !a}
+  deriving (Functor, Show)
 
 -- | Returns the rendered output and the
 -- matching score for a pattern and a text.
 -- Two examples are given below:
 --
--- >>> match "fnt" "infinite" "" "" id True
--- Just ("infinite",3)
+-- >>> match "fnt" "infinite"
+-- Just 3
 --
--- >>> match "hsk" ("Haskell",1995) "<" ">" fst False
--- Just ("<h>a<s><k>ell",5)
+-- >>> match "hsk" "Haskell"
+-- Just 5
 --
 {-# INLINABLE match #-}
 
-match :: (T.TextualMonoid s)
-      => s        -- ^ Pattern in lowercase except for first character
-      -> t        -- ^ The value containing the text to search in.
-      -> s        -- ^ The text to add before each match.
-      -> s        -- ^ The text to add after each match.
-      -> (t -> s) -- ^ The function to extract the text from the container.
-      -> Maybe (Fuzzy t s) -- ^ The original value, rendered string and score.
-match pattern t pre post extract =
-    if null pat then Just (Fuzzy t result totalScore) else Nothing
+match :: T.Text    -- ^ Pattern in lowercase except for first character
+      -> T.Text    -- ^ The text to search in.
+      -> Maybe Int -- ^ The score
+match (T.Text pArr pOff pLen) (T.Text sArr sOff sLen) = go 0 1 pOff sOff
```

Comment:
Back in the day I tried the specialising trick but it didn't make a difference. GHC was producing the same or worse code